libphobos/m4/druntime/libraries.m4

```diff
@@ -6,27 +6,135 @@
                  [specify linker option for the system thread library (default: autodetect)]),
-  [d_thread_lib=$enableval],[d_thread_lib=""])
+  [enable_thread_lib=$enableval],[enable_thread_lib="yes"])
+
+  AS_IF([test "x$enable_thread_lib" = "xyes"], [
+    AC_SEARCH_LIBS([pthread_create], [pthread])
+  ], [
+    AS_IF([test "x$enable_thread_lib" = "xno"], [
+      AC_MSG_CHECKING([for thread library])
+      AC_MSG_RESULT([disabled])
+      ], [
+      AC_CHECK_LIB([$enable_thread_lib], [pthread_create], [
+        LIBS="-l$enable_thread_lib $LIBS"
+        ], [
+        AC_MSG_ERROR([Thread library was not found])])
+        ])
+      ])
+  ])
 ])
 
 
 # DRUNTIME_LIBRARIES_ZLIB
 # -----------------------
 # Allow specifying whether to use the system zlib or
 # compiling the zlib included in GCC. Define
-# DRUNTIME_ZLIB_SYSTEM conditional
+# DRUNTIME_ZLIB_SYSTEM conditional and add zlib to
+# LIBS if necessary.
 AC_DEFUN([DRUNTIME_LIBRARIES_ZLIB],
 [
-  dnl switch between system zlib and gcc zlib
-  AC_ARG_WITH(system-zlib,
-    AS_HELP_STRING([--with-system-zlib],
-                   [use installed libz (default: no)]),
-    [system_zlib=yes],[system_zlib=no])
-  AM_CONDITIONAL([DRUNTIME_ZLIB_SYSTEM], [test "$system_zlib" = yes])
+  AC_ARG_WITH(target-system-zlib,
+    AS_HELP_STRING([--with-target-system-zlib],
+                   [use installed libz (default: no)]))
+
+  system_zlib=false
+  AS_IF([test "x$with_target_system_zlib" = "xyes"], [
+    AC_CHECK_LIB([z], [deflate], [
+      LIBS="-lz $LIBS"
+      system_zlib=yes
+      ], [
+      AC_MSG_ERROR([System zlib was not found])])
+  ], [
+    AC_MSG_CHECKING([for zlib])
+    AC_MSG_RESULT([just compiled])
+  ])
+
+  AM_CONDITIONAL([DRUNTIME_ZLIB_SYSTEM], [test "$with_target_system_zlib" = yes])
+])
+
+# DRUNTIME_LIBRARIES_C
+# -------------------------
+# Allow specifying the libc library to link with or autodetect
+# Add libc library to LIBS if necessary.
+AC_DEFUN([DRUNTIME_LIBRARIES_C],
+[
+AC_ARG_ENABLE(c-lib,
+  AC_HELP_STRING([--enable-c-lib=<arg>],
+                 [specify linker option for the system C library (default: autodetect)]),
+  [enable_c_lib=$enableval],[enable_c_lib="yes"])
+
+  AS_IF([test "x$enable_c_lib" = "xyes"], [
+    AC_SEARCH_LIBS([malloc], [c])
+  ], [
+    AS_IF([test "x$enable_c_lib" = "xno"], [
+      AC_MSG_CHECKING([for C library])
+      AC_MSG_RESULT([disabled])
+      ], [
+      AC_CHECK_LIB([$enable_c_lib], [malloc], [
+        LIBS="-l$enable_c_lib $LIBS"
+        ], [
+        AC_MSG_ERROR([C library was not found])])
+        ])
+      ])
+  ])
+])
+
+# DRUNTIME_LIBRARIES_RT
+# -------------------------
+# Allow specifying the rt library to link with or autodetect
+# Add rt library to LIBS if necessary.
+AC_DEFUN([DRUNTIME_LIBRARIES_RT],
+[
+AC_ARG_ENABLE(rt-lib,
+  AC_HELP_STRING([--enable-rt-lib=<arg>],
+                 [specify linker option for the system rt library (default: autodetect)]),
```

Comment:
You're right. I've updated the macros to fail if autodetection of a library fails.

Explicit disabling (--disable-thread-lib) can still be useful when our configure scripts are not working 100% correctly (e.g. checking for posix functions only when druntime needs a windows lib/function).